providers/common/ciphers/aes_gcm_s390x.inc

```diff
@@ -0,0 +1,327 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * IBM S390X support
+ */
+# include "s390x_arch.h"
+
+/* Most modes of operation need km for partial block processing. */
+#  define S390X_aes_128_CAPABLE (OPENSSL_s390xcap_P.km[0] &  \
+                                S390X_CAPBIT(S390X_AES_128))
+#  define S390X_aes_192_CAPABLE (OPENSSL_s390xcap_P.km[0] &  \
+                                S390X_CAPBIT(S390X_AES_192))
+#  define S390X_aes_256_CAPABLE (OPENSSL_s390xcap_P.km[0] &  \
+                                S390X_CAPBIT(S390X_AES_256))
+
+/* Convert key size to function code: [16,24,32] -> [18,19,20]. */
+# define S390X_AES_FC(keylen)  (S390X_AES_128 + ((((keylen) << 3) - 128) >> 6))
+
+# define S390X_aes_128_gcm_CAPABLE (S390X_aes_128_CAPABLE &&        \
+                                    (OPENSSL_s390xcap_P.kma[0] &    \
+                                     S390X_CAPBIT(S390X_AES_128)))
+# define S390X_aes_192_gcm_CAPABLE (S390X_aes_192_CAPABLE &&        \
+                                    (OPENSSL_s390xcap_P.kma[0] &    \
+                                     S390X_CAPBIT(S390X_AES_192)))
+# define S390X_aes_256_gcm_CAPABLE (S390X_aes_256_CAPABLE &&        \
+                                    (OPENSSL_s390xcap_P.kma[0] &    \
+                                     S390X_CAPBIT(S390X_AES_256)))
+
+/* iv + padding length for iv lengths != 12 */
+# define S390X_gcm_ivpadlen(i)  ((((i) + 15) >> 4 << 4) + 16)
+
+static int s390x_aes_gcm_init_key(PROV_AES_GCM_KEY *ctx,
+                                  const unsigned char *key, size_t keylen)
+{
+    ctx->key_set = 1;
+    memcpy(&ctx->plat.s390x.param.kma.k, key, keylen);
+    ctx->plat.s390x.fc = S390X_AES_FC(keylen);
+    if (ctx->enc)
+        ctx->plat.s390x.fc |= S390X_DECRYPT;
+    return 1;
+}
+
+/*-
+ * Initialize context structure. Code is big-endian.
+ */
+static void s390x_aes_gcm_setiv(PROV_AES_GCM_KEY *ctx,
+                                const unsigned char *iv,
+                                size_t ivlen)
+{
+    S390X_KMA_PARAMS *kma = &ctx->plat.s390x.param.kma;
+
+    kma->t.g[0] = 0;
+    kma->t.g[1] = 0;
+    kma->tpcl = 0;
+    kma->taadl = 0;
+    ctx->plat.s390x.mreslen = 0;
+    ctx->plat.s390x.areslen = 0;
+    ctx->plat.s390x.kreslen = 0;
+
+    if (ivlen == 12) {
+        memcpy(&kma->j0, iv, ivlen);
+        kma->j0.w[3] = 1;
+        kma->cv.w = 1;
+    } else {
+        size_t len = S390X_gcm_ivpadlen(ctx->ivlen);
+
+        if (len >= sizeof(ctx->iv))
+            return 0;
+        /*
+         * The IV length needs to be zero padded to be a multiple of 16 bytes
+         * followed by 8 bytes of zeros and 8 bytes for the IV length.
+         * The GHASH of this value can then be calculated.
+         */
+        memcpy(ctx->iv, iv, ctx->ivlen);
+        memset(ctx->iv + ctx->ivlen, 0, len - ivlen);
+        *((unsigned long long *)(ctx->iv + len - 8)) = ivlen << 3;
```

Comment:
I know this was just copied from legacy code, however strict-warnings build complains in this place ..
```
In file included from providers/common/ciphers/aes_gcm_generic.c:211:0:
providers/common/ciphers/aes_gcm_s390x.inc: In function 's390x_aes_gcm_setiv':
providers/common/ciphers/aes_gcm_s390x.inc:83:9: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
         *((unsigned long long *)(ctx->iv + len - 8)) = ivlen << 3;
         ^
```

How about :
```
diff --git a/providers/common/ciphers/aes_gcm_s390x.inc b/providers/common/ciphers/aes_gcm_s390x.inc
index f3c40eb55c..056e77cbfc 100644
--- a/providers/common/ciphers/aes_gcm_s390x.inc
+++ b/providers/common/ciphers/aes_gcm_s390x.inc
@@ -69,6 +69,7 @@ static void s390x_aes_gcm_setiv(PROV_AES_GCM_KEY *ctx,
         kma->j0.w[3] = 1;
         kma->cv.w = 1;
     } else {
+        unsigned long long ivbits = ivlen << 3; /* use preferred u64 type here .. */
         size_t len = S390X_gcm_ivpadlen(ctx->ivlen);
 
         if (len >= sizeof(ctx->iv))
@@ -80,7 +81,8 @@ static void s390x_aes_gcm_setiv(PROV_AES_GCM_KEY *ctx,
          */
         memcpy(ctx->iv, iv, ctx->ivlen);
         memset(ctx->iv + ctx->ivlen, 0, len - ivlen);
-        *((unsigned long long *)(ctx->iv + len - 8)) = ivlen << 3;
+        memcpy(ctx->iv + len - sizeof(ivbits), &ivbits, sizeof(ivbits));
         /*
          * Calculate the ghash of the iv - the result is stored into the tag
          * param.

```